DOCSP-35937 aggregations builder 4.3 (#2876)
jenssegers/laravel-mongodb: docs/fundamentals/aggregation-builder.txt

.. _laravel-aggregation-builder:

===================
Aggregation Builder
===================

.. facet::
   :name: genre
   :values: tutorial

.. meta::
   :keywords: code example, pipeline, expression

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to perform aggregations and construct
pipelines by using the {+odm-short+} aggregation builder. The aggregation
builder lets you use a typesafe syntax to construct a MongoDB
**aggregation pipeline**.

An aggregation pipeline is a data processing pipeline that takes documents

> as input, sequentially performs transformations and computations on the data,

Reviewer: ```suggestion
An aggregation pipeline is a data processing pipeline that
sequentially performs transformations and computations on input data,
```

The {+odm-short+} aggregation builder lets you build aggregation stages and
aggregation pipelines. The following sections show examples of how to use the
aggregation builder to create the stages of an aggregation pipeline:

- :ref:`laravel-build-aggregation`
- :ref:`laravel-create-custom-operator-factory`

.. tip::

   The aggregation builder feature is available only in {+odm-short+} versions
   4.3 and later. To learn more about running aggregations without using the
   aggregation builder, see :ref:`laravel-query-builder-aggregations` in the
   Query Builder guide.

.. _laravel-build-aggregation:

Create an Aggregation Pipeline
------------------------------

To start an aggregation pipeline, call the ``Model::aggregate()`` method.
Then, chain the aggregation stage methods in the sequence you want them to
run.

The aggregation builder includes the following namespaces that you can import
to build aggregation stages:

- ``MongoDB\Builder\Accumulator``
- ``MongoDB\Builder\Expression``
- ``MongoDB\Builder\Query``
- ``MongoDB\Builder\Type```

.. tip::

   To learn more about builder classes, see the `mongodb/mongodb-php-builder <https://github.com/mongodb/mongo-php-builder/>`__

>    GitHub repository.

Reviewer 2: You should mention that the additionnal package `mongodb/builder` must be required in `composer.json` or by running `composer require mongodb/builder`

Author: Good catch. Will add a section on this.

This section features the following examples of common aggregation stages and

> of an aggregation pipeline:

Reviewer: ```suggestion
This section features the following examples, which demonstrate how to use common
aggregation stages and combine stages to make an aggregation pipeline:
```

the {+server-docs-name+}.

Sample Documents
~~~~~~~~~~~~~~~~

The following examples run aggregation pipelines on a collection represented
by the ``User`` model. You can add the sample data by running the following
``insert()`` method:

.. literalinclude:: /includes/fundamentals/aggregation/AggregationsBuilderTest.php
      :language: php
      :dedent:
      :start-after: begin aggregation builder sample data
      :end-before: end aggregation builder sample data

.. _laravel-aggregation-match-stage-example:

Match Stage Example
~~~~~~~~~~~~~~~~~~~

You can chain the ``match()`` method to your aggregation pipeline to specify
a query filter. If you omit this stage, the ``aggregate()`` method outputs
all the documents in the model's collection for the following stage.

This aggregation stage is often placed first to retrieve the data by using
available indexes and reduce the amount of data the subsequent stages process.

.. tip::

   If you omit the ``match()`` method, the aggregation pipeline matches all
   documents in the collection that correspond to the model before other
   aggregation stages.

This example constructs a query filter for a **match** aggregation stage by

> using the ``Query`` builder. The match stage includes the the following

Reviewer 2: The full namespace would be helpful. Unfortunately it's impossible to make it appear in the example without making it extremely verbose.

```suggestion
using the ``MongoDB\Builder\Query`` builder. The match stage includes the the following
```

criteria:

- Returns results that match either of the query filters by using the

>   ``Query::or()`` function

Reviewer 2: Maybe link to the [`$or` operator page](https://www.mongodb.com/docs/manual/reference/operator/query/or/), that's a great way to explore other operators.

Author: Good idea, adding a link at the end of the example.

You can chain the ``group()`` method to your aggregation pipeline to modify the
structure of the data by performing calculations and grouping it by common
field values.

This aggregation stage is often placed immediately after a match stage to
reduce the data subsequent stages process.

This example uses the ``MongoDB\Builder\Expression`` builder to define the group keys in a
**group** aggregation stage. The group stage specifies the following grouping
behavior:

- Sets the value of the group key, represented by the ``_id`` field, to the
  field value defined by the ``Expression`` builder
- References the document values in the ``occupation`` field by calling the
  ``Expression::fieldPath()`` function

Click the :guilabel:`{+code-output-label+}` button to see the documents
returned by running the code:

.. io-code-block::

   .. input:: /includes/fundamentals/aggregation/AggregationsBuilderTest.php
      :language: php
      :dedent:
      :start-after: begin aggregation group stage
      :end-before: end aggregation group stage

   .. output::
      :language: json
      :visible: false

      [
        { "_id": "engineer" },
        { "_id": "designer" }
      ]

.. tip::

>    This example stage performs a similar grouping to an equivalent
>    ``distinct()`` query builder method. To learn more about the ``distinct()``
>    method, see the :ref:`laravel-distinct-usage` usage example.

Reviewer: ```suggestion
   This example stage performs a similar task as the
   ``distinct()`` query builder method. To learn more about the ``distinct()``
   method, see the :ref:`laravel-distinct-usage` usage example.
```

            }
          }
        },
        {
          "_id": ...,
          "name": "Ellis Lee",
          "occupation": "designer",
          "birthday": {
            "$date": {
              "$numberLong": "834019200000"
            }
          }
        },
        {
          "_id": ...,
          "name": "Eliud Nkosana",
          "occupation": "engineer",
          "birthday": {
            "$date": {
              "$numberLong": "449884800000"
            }
          }
        },
        {
          "_id": ...,
          "name": "Bran Steafan",
          "occupation": "engineer",
          "birthday": {
            "$date": {
              "$numberLong": "894326400000"
            }
          }
        },
        {
          "_id": ...,
          "name": "Alda Gröndal",
          "occupation": "engineer",
          "birthday": {
            "$date": {
              "$numberLong": "1009843200000"
            }
          }
        }
      ]

.. _laravel-aggregation-project-stage-example:

Project Stage Example
~~~~~~~~~~~~~~~~~~~~~

You can chain the ``project()`` method to your aggregation pipeline to specify
which fields from the documents to display by this stage.

> To specify fields to include, pass the name of a field and a truthy value,

Reviewer: Q: could including words like "truthy" and "falsy" confuse readers?

Author: I think it's a relatively common term and not obscure jargon. I provided examples in the qualifying phrase immediately after. Let me know if you had something else in mind for clarifying this.

This example shows how to use the ``project()`` method aggregation stage to
include only the ``name`` field and exclude all other fields from the output.
Click the :guilabel:`{+code-output-label+}` button to see the data returned by
running the code:

.. io-code-block::

   .. input:: /includes/fundamentals/aggregation/AggregationsBuilderTest.php
      :language: php
      :dedent:
      :start-after: begin aggregation project stage
      :end-before: end aggregation project stage

   .. output::
      :language: json
      :visible: false

      [
        { "name": "Alda Gröndal" },
        { "name": "Francois Soma" },
        { "name": "Janet Doe" },
        { "name": "Eliud Nkosana" },
        { "name": "Bran Steafan" },
        { "name": "Ellis Lee" }
      ]

.. _laravel-aggregation-pipeline-example:

Aggregation Pipeline Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This aggregation pipeline example chains multiple stages. Each stage runs
on the output retrieved from each preceding stage. In this example, the
stages perform the following operations sequentially:

- Match all documents in the collection since the ``match()`` stage

>   is omitted.

Reviewer: Q: is this line necessary? It could be confusing to mention things that the pipeline does NOT include

Author: I thought it would be good to reinforce the information provided in the admonition in the Match Stage Example section. I think it is relevant to the description of what the pipeline does and might cause readers to question what documents are part of the pipeline if omitted.

Let me know if you think it would be better to omit it.

Reviewer: I think it could work better as a note, potentially, but I am ok with leaving as is too!

Author: Good idea, will move it to an admonition.

> - Add the ``year`` field to the documents and set the value to the year
>   extracted from the ``birthday`` field.

Reviewer: S: could be more descriptive
```suggestion
- Add the ``birth_year`` field to the documents and set the value to the year
  extracted from the ``birthday`` field.
```

Reviewer: If you change this, change the `year_avg` field to `birth_year_avg`

Author: Thanks, will update the example to use birth_year and birth_year_avg field names.

.. _laravel-aggregation-builder:

===================
Aggregation Builder
===================

.. facet::
   :name: genre
   :values: tutorial

.. meta::
   :keywords: code example, pipeline, expression

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to perform aggregations and construct
pipelines by using the {+odm-short+} aggregation builder. The aggregation

> builder lets you use a typesafe syntax to construct a MongoDB

Reviewer: ```suggestion
builder lets you use a type-safe syntax to construct a MongoDB
```

**aggregation pipeline**.

An aggregation pipeline is a data processing pipeline that sequentially

> performs transformations and computations on input data, and outputs the
> results as a new set of documents.

Reviewer: ```suggestion
performs transformations and computations on input data, then outputs the
results as a new document or set of documents.
```

> An aggregation pipeline is composed of **aggregation stages**, which are
> operators that process input data and output data that the next stage uses as
> its input.

Reviewer: ```suggestion
An aggregation pipeline is composed of **aggregation stages**, which use
operators to process input data and produce data that the next stage uses as
its input.
```

Click the :guilabel:`{+code-output-label+}` button to see the documents
returned by running the code:

.. io-code-block::

   .. input:: /includes/fundamentals/aggregation/AggregationsBuilderTest.php
      :language: php
      :dedent:
      :start-after: begin aggregation group stage
      :end-before: end aggregation group stage

   .. output::
      :language: json
      :visible: false

      [
        { "_id": "engineer" },
        { "_id": "designer" }
      ]

.. tip::

   This example stage performs a similar task as the ``distinct()`` query
   builder method. To learn more about the ``distinct()`` method, see the
   :ref:`laravel-distinct-usage` usage example.

.. _laravel-aggregation-sort-stage-example:

Sort Stage Example
~~~~~~~~~~~~~~~~~~

You can chain the ``sort()`` method to your aggregation pipeline to specify
the documents' output order.

You can add this aggregation stage anywhere in the pipeline. It is often
placed after the group stage since it can depend on the grouped data. We
recommend placing the sort stage as late as possible in the pipeline to limit
the data it processes.

To specify an ascending sort, set the field value to the ``Sort::Asc`` enum.

> To specify a descending sort, set the field value to the ``Sort::Desc`` enum.

Reviewer: S; combine into one sentence to shorten this section/reduce breaks

running the code:

.. io-code-block::

   .. input:: /includes/fundamentals/aggregation/AggregationsBuilderTest.php
      :language: php
      :dedent:
      :start-after: begin pipeline example
      :end-before: end pipeline example

   .. output::
      :language: json
      :visible: false

      [
        {
          "birth_year_avg": 1991.5,
          "profession": "designer"
        },
        {
          "birth_year_avg": 1995.5,
          "profession": "engineer"
        }
      ]

.. _laravel-create-custom-operator-factory:

Create a Custom Operator Factory
--------------------------------

When using the aggregation builder to create an aggregation pipeline, you
can define operations or stages in a **custom operator factory**. A custom
operator factory is a function that returns expressions or stages of an
aggregation pipeline. You can create these functions to improve code
readability and reuse.

This example shows how to create and use a custom operator factory that
returns expressions that extract the year from a specified date field.

The following function accepts the name of a field that contains a date

> and returns an expression that extracts the year from the date:

Reviewer: S: I think this section should provide an example of a functionality that is not directly achievable from one aggregation operator, `$year` in this case. For example, could this example extract the year and add a fixed number to calculate something like the year they turned 18? Or a field to tell if they were born in the 20th century or 21st century or something like that.

Author: Thanks -- I think that's a great idea and I had thought of other ideas. However, I intentionally decided on something less complex to focus on documenting the custom operator factory usage.

.. _laravel-aggregation-builder:

===================
Aggregation Builder
===================

.. facet::
   :name: genre
   :values: tutorial

.. meta::
   :keywords: code example, pipeline, expression

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to perform aggregations and construct
pipelines by using the {+odm-short+} aggregation builder. The aggregation
builder lets you use a type-safe syntax to construct a MongoDB
**aggregation pipeline**.

An aggregation pipeline is a data processing pipeline that sequentially

> performs transformations and computations on input data, then outputs the

Reviewer 2: "input data" sounds like data that you send with the command. It think this working could be better: "data from the MongoDB database"
